### PR TITLE
[DNM] sql,sqlccl: allow backups to be read in SQL

### PR DESCRIPTION
**Pull request by danhhz**
A new table_ref is added that allows a user to specify a single table
and set of backups (full or full + incremental). This is turned into a
planDataSource with a new field (BackupSource) filled in on the
TableDescriptor. AS OF SYSTEM TIME is also supported.

Examples:
  TABLE EXPERIMENTAL_BACKUP_SOURCE db.table FROM ('backup1', 'backup2')
  SELECT * FROM [TABLE EXPERIMENTAL_BACKUP_SOURCE ...] AS OF SYSTEM TIME foo

MultiRowFetcher now switches on the presence of BackupSource on a
TableDescriptor to use a different implementation of the KVFetcher
interface. Instead of fetching via our KV layer, the sstables of the
backup are downloaded and iterated in-memory.

Unfortunately, it's not possible to use placeholders for the backup
paths (which is really helpful for BACKUP/RESTORE as you don't have to
worry about escaping the strings) for the same reason that table name
can't be a placeholder in a normal SELECT.

TODO
- test for hole in backup descs
- test that it works with dist sql
- test that it's lazily downloading files
- tests for various error conditions

Release note (enterprise change): Add an experimental SQL feature for
SELECTing data from backups.

Reviewers: there is obviously a lot of godoc, polish, and testing to be done here, but I wanted to get it out for early review on the approach. In particular, the syntax is pretty awful right now and I'd love some bike-shedding on it (though after playing with various options, I'm fairly sure that making it a `tableExpr` is the right thing). Also, what other edge cases should I be testing? I listed a few as a TODO above, but I'm probably missing a bunch and my "tests for various error conditions" is hand-wavey.

cc @jordanlewis for thoughts on the sql side of this

### CODE REVIEW

**Review by jordanlewis — COMMENTED** [review]
This is super cool!! I think you need to update the PR description to match the commit message.

Syntax bikeshed: What about this option?

```
SELECT * FROM EXPERIMENTAL_BACKUP_SOURCE('%s', '%s) data.bank
```

File: pkg/sql/sqlbase/multirowfetcher.go [hunk 1]
```diff
@@ -36,9 +37,10 @@ import (
 // this to avoid using log.V in the hot path.
 const debugRowFetch = false
 
-type kvFetcher interface {
-	nextKV(ctx context.Context) (bool, roachpb.KeyValue, error)
-	getRangesInfo() []roachpb.RangeInfo
+// KVFetcher ...
```
Comment (jordanlewis) [inline pkg/sql/sqlbase/multirowfetcher.go]: I'm sure you know about this already, but just a reminder to fix this doc. Could add a TODO(dan) BEFORE MERGE :)

**Comment by dt** [comment]
just to contribute some more paint sample swatches to this bike shed, my 2¢ would also be (if i'm reading jordan's suggestion correctly) to make the URIs look like args rather than `FROM` a top-level list. I might even try to put it all inside the source's braces, but i'm not sure how to do that in a way that looks good. 

```
SELECT bank.id, bank.balance, oldbank.balance
FROM bank
LEFT JOIN BACKUP_SOURCE(data.bank FROM 'blah', 'blah') AS oldbank ON bank.id = oldbank.id
```

---

Reviewed 19 of 20 files at r1.
Review status: 19 of 20 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup_source.go, line 54 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20664#-L1xMGqAoeZzktGeFDWB:-L1xMGqAoeZzktGeFDWC:b-qizais) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2124f759d75c8bd373ee84d1317d5cb665286c9/pkg/ccl/sqlccl/backup_source.go#L54)):*
> ```Go
> 	tn := &tree.TableName{
> 		TableName: tree.Name(tableDesc.Name),
> 		// TODO(dan): BEFORE MERGE Fill in Database.
> ```

:thumbsup: as mentioned below, i'd make sure that tests hit lots of qualified name cases, and preferably, like the example above, try to collide with real db and table names, use aliases, etc as well.

---

*[pkg/ccl/sqlccl/backup_source.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20664#-L1xLU7QtQQcRdo52PSY:-L1xLU7QtQQcRdo52PSZ:b-ocz5y7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2124f759d75c8bd373ee84d1317d5cb665286c9/pkg/ccl/sqlccl/backup_source.go#L76)):*
> ```Go
> 
> 	for _, uri := range tableDesc.BackupSource.Uris {
> 		exportStore, err := exportStorageFromURI(ctx, uri, params.Settings)
> ```

Do we need to call `Close` ourselves in the non-nil error return cases to close whatever export stores we've managed to open? 

---

*[pkg/ccl/sqlccl/backup_source.go, line 90 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20664#-L1xbvQAVonVTPbLDcKB:-L1xbvQAVonVTPbLDcKC:b-t5ugkr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2124f759d75c8bd373ee84d1317d5cb665286c9/pkg/ccl/sqlccl/backup_source.go#L90)):*
> ```Go
> 	}
> 
> 	// TODO(dan): Avoid loading all the backup descriptors a second time.
> ```

I have an open PR that had this same complaint so I switched to loading the descriptors separately and passing them in to it, at which point there wasn't anything left in `ValidatePreviousBackups` so I just eliminated it in favor of just calling `makeImportSpans` directly.

---

*[pkg/ccl/sqlccl/backup_test.go, line 313 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20664#-L1xGINvkU0O9TS2vI4n:-L1xGINwBNEO-Ftt3QPw:b-rskodc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2124f759d75c8bd373ee84d1317d5cb665286c9/pkg/ccl/sqlccl/backup_test.go#L313)):*
> ```Go
> 		sqlDBRestore := sqlutils.MakeSQLRunner(tcRestore.Conns[0])
> 
> 		sqlDBRestore.CheckQueryResults(t, fmt.Sprintf(`TABLE EXPERIMENTAL_BACKUP_SOURCE data.bank FROM ('%s')`, dest), data)
> ```

TIL: `TABLE foo`.

Do we want to validate that we get an expected error on attempts to update/insert on these fake tables? Or is that already tested in some general datasource handler elsewhere?

---

*[pkg/ccl/sqlccl/backup_test.go, line 1768 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20664#-L1xHIDTZuQuAkpTc1UO:-L1xHIDTZuQuAkpTc1UP:bt91kz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2124f759d75c8bd373ee84d1317d5cb665286c9/pkg/ccl/sqlccl/backup_test.go#L1768)):*
> ```Go
> 			)
> 			sqlDB.CheckQueryResults(t,
> 				fmt.Sprintf(`SELECT * FROM EXPERIMENTAL_BACKUP_SOURCE data.bank FROM ('%s', '%s') AS OF SYSTEM TIME %s`, fullBackup, incBackup, timestamp),
> ```

Would be good to include a test that uses some qualified names and aliases, e.g.  `SELECT bank.id` or `SELECT data.bank.*` and maybe aliased too, e.g. `SELECT  foo.id FROM ... AS foo` 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20664#-:-L1xI5pLS8hwBr03WKWY:bmprz7m)*

**Comment by danhhz** [comment]
This is two releases out of date at this point. I suspect resolving the merge conflicts would be pretty close to rewriting it. Bummer to lose the review comments here, but this has been the one thing sitting in my outstanding PRs for a year now and it bugs me, so I'm going to close it. : - )

Made a note that there were review comments here in https://github.com/cockroachdb/cockroach/pull/20664

**Pull request closed without merging**